### PR TITLE
[SPARK-36456][CORE][SQL][SS] Clean up compilation warnings related to `method closeQuietly in class IOUtils is deprecated`

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
There are some compilation warnings related to `method closeQuietly in class IOUtils is deprecated`,  `Apache commons-io` suggests that we should use the `try-with-resources` statement or handle suppressed exceptions manually.

The main change of this pr is replace  `o.a.commons.io.IOUtils.closeQuietly` with `o.a.s.network.util.JavaUtils.closeQuietly` directly because all original logic is suppressing `IOException`.
 

### Why are the changes needed?
Clean up compilation warnings related to `method closeQuietly in class IOUtils is deprecated`

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?

- Pass the Jenkins or GitHub Action

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 1]
```diff
@@ -2657,6 +2657,24 @@ private[spark] object Utils extends Logging {
...
   }
 
+  /**
+   * Closes the given Closeable as a null-safe operation while consuming IOException
+   * by the given consumer.
+   *
+   * @param closeable The resource to close, may be null.
+   * @param consumer Consumes the IOException thrown by Closeable#close().
+   */
+  def closeQuietly(closeable: Closeable, consumer: IOException => Unit = null): Unit = {
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/util/Utils.scala:2667]: Is the logic identical, @LuciferYang ?

Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/util/Utils.scala:2667]: The logic is the same, but `IOUtils.closeQuietly ` has been identified as `@Deprecated `.

And I found a similar method encapsulation in `o.a.s.network.util.JavaUtils.java`, should we reuse it?

https://github.com/apache/spark/blob/c040593fb432f75f7e22e3d32a6979142a125375/common/network-common/src/main/java/org/apache/spark/network/util/JavaUtils.java#L50-L58

Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/util/Utils.scala:2667]: Thank you for the confirmation. Yes, we had better reuse it.

**Comment by dongjoon-hyun** [comment]
After making this to reuse it, we can merge this.

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala [hunk 2]
```diff
@@ -29,7 +29,6 @@ import scala.collection.mutable.{ArrayBuffer, HashMap, HashSet, Queue}
 import scala.util.{Failure, Success}
 
 import io.netty.util.internal.OutOfDirectMemoryError
-import org.apache.commons.io.IOUtils
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:32]: BTW, do you know why it's deprecated?

File: core/src/main/scala/org/apache/spark/util/Utils.scala (hunk 1 above)
Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/util/Utils.scala:2667]: ok

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala (hunk 2 above)
Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:32]:
```
@deprecated As of 2.6 deprecated without replacement. Please use the try-with-resources statement or handle
suppressed exceptions manually.
```

All our original logic is suppressing `IOException`, so I think we can reuse `JavaUtils. closeQuietly` directly If all modules can access it.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. (Pending CIs).
Thank you, @LuciferYang .

**Review by HyukjinKwon — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 3.3.
Thank you, @LuciferYang and @HyukjinKwon .

**Comment by LuciferYang** [comment]
thx @dongjoon-hyun @HyukjinKwon

**Comment by HeartSaVioR** [comment]
Hello,

`JavaUtils.closeQuietly` looks to be slightly different from `IOUtils.closeQuietly` in default behavior. Swallowing IOException is same, but the former logs it as ERROR while the latter doesn't log by default. While it might be arguable which one is the right behavior, there seems to be some case we know IOException could happen but no big deal, like cancelling delta/metadata file in CancellableFSDataOutputStream, which we don't like to produce error log message.

```
[info] 12:24:05.252 ERROR org.apache.spark.network.util.JavaUtils: IOException should not have been thrown.
[info] java.nio.channels.ClosedChannelException
[info] 	at org.apache.hadoop.fs.ChecksumFs$ChecksumFSOutputSummer.checkClosed(ChecksumFs.java:390)
[info] 	at org.apache.hadoop.fs.FSOutputSummer.write(FSOutputSummer.java:106)
[info] 	at org.apache.hadoop.fs.FSDataOutputStream$PositionCache.write(FSDataOutputStream.java:62)
[info] 	at java.io.DataOutputStream.write(DataOutputStream.java:107)
[info] 	at org.apache.hadoop.fs.FSDataOutputStream$PositionCache.write(FSDataOutputStream.java:62)
[info] 	at java.io.DataOutputStream.write(DataOutputStream.java:107)
[info] 	at com.github.luben.zstd.ZstdOutputStreamNoFinalizer.flush(ZstdOutputStreamNoFinalizer.java:249)
[info] 	at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:141)
[info] 	at java.io.DataOutputStream.flush(DataOutputStream.java:123)
[info] 	at java.io.FilterOutputStream.close(FilterOutputStream.java:158)
[info] 	at org.apache.spark.network.util.JavaUtils.closeQuietly(JavaUtils.java:53)
[info] 	at org.apache.spark.sql.execution.streaming.state.HDFSBackedStateStoreProvider.org$apache$spark$sql$execution$streaming$state$HDFSBackedStateStoreProvider$$cancelDeltaFile(HDFSBackedStateStoreProvider.scala:545)
[info] 	at org.apache.spark.sql.execution.streaming.state.HDFSBackedStateStoreProvider$HDFSBackedStateStore.abort(HDFSBackedStateStoreProvider.scala:158)
[info] 	at org.apache.spark.sql.execution.benchmark.StateStoreBenchmark$.$anonfun$runBenchmarkSuite$3(StateStoreBenchmark.scala:95)
[info] 	at org.apache.spark.sql.execution.benchmark.StateStoreBenchmark$.$anonfun$runBenchmarkSuite$3$adapted(StateStoreBenchmark.scala:84)
[info] 	at org.apache.spark.benchmark.Benchmark.measure(Benchmark.scala:149)
[info] 	at org.apache.spark.benchmark.Benchmark.$anonfun$run$1(Benchmark.scala:106)
[info] 	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:286)
[info] 	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
[info] 	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
[info] 	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
[info] 	at scala.collection.TraversableLike.map(TraversableLike.scala:286)
[info] 	at scala.collection.TraversableLike.map$(TraversableLike.scala:279)
[info] 	at scala.collection.AbstractTraversable.map(Traversable.scala:108)
[info] 	at org.apache.spark.benchmark.Benchmark.run(Benchmark.scala:104)
[info] 	at org.apache.spark.sql.execution.benchmark.StateStoreBenchmark$.$anonfun$runBenchmarkSuite$1(StateStoreBenchmark.scala:112)
[info] 	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[info] 	at org.apache.spark.benchmark.BenchmarkBase.runBenchmark(BenchmarkBase.scala:42)
[info] 	at org.apache.spark.sql.execution.benchmark.StateStoreBenchmark$.runBenchmarkSuite(StateStoreBenchmark.scala:58)
[info] 	at org.apache.spark.benchmark.BenchmarkBase.main(BenchmarkBase.scala:72)
[info] 	at org.apache.spark.sql.execution.benchmark.StateStoreBenchmark.main(StateStoreBenchmark.scala)
[info] 	Suppressed: java.nio.channels.ClosedChannelException
[info] 		at org.apache.hadoop.fs.ChecksumFs$ChecksumFSOutputSummer.checkClosed(ChecksumFs.java:390)
[info] 		at org.apache.hadoop.fs.FSOutputSummer.write(FSOutputSummer.java:106)
[info] 		at org.apache.hadoop.fs.FSDataOutputStream$PositionCache.write(FSDataOutputStream.java:62)
[info] 		at java.io.DataOutputStream.write(DataOutputStream.java:107)
[info] 		at org.apache.hadoop.fs.FSDataOutputStream$PositionCache.write(FSDataOutputStream.java:62)
[info] 		at java.io.DataOutputStream.write(DataOutputStream.java:107)
[info] 		at com.github.luben.zstd.ZstdOutputStreamNoFinalizer.flush(ZstdOutputStreamNoFinalizer.java:249)
[info] 		at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:141)
[info] 		at java.io.FilterOutputStream.close(FilterOutputStream.java:158)
[info] 		at java.io.FilterOutputStream.close(FilterOutputStream.java:159)
[info] 		... 21 more
[info] 		Suppressed: java.nio.channels.ClosedChannelException
[info] 			at org.apache.hadoop.fs.ChecksumFs$ChecksumFSOutputSummer.checkClosed(ChecksumFs.java:390)
[info] 			at org.apache.hadoop.fs.FSOutputSummer.write(FSOutputSummer.java:106)
[info] 			at org.apache.hadoop.fs.FSDataOutputStream$PositionCache.write(FSDataOutputStream.java:62)
[info] 			at java.io.DataOutputStream.write(DataOutputStream.java:107)
[info] 			at org.apache.hadoop.fs.FSDataOutputStream$PositionCache.write(FSDataOutputStream.java:62)
[info] 			at java.io.DataOutputStream.write(DataOutputStream.java:107)
[info] 			at com.github.luben.zstd.ZstdOutputStreamNoFinalizer.close(ZstdOutputStreamNoFinalizer.java:279)
[info] 			at com.github.luben.zstd.ZstdOutputStreamNoFinalizer.close(ZstdOutputStreamNoFinalizer.java:258)
[info] 			at java.io.FilterOutputStream.close(FilterOutputStream.java:159)
[info] 			... 22 more
```

Furthermore, looks like commons-io community decided to retain the method.

https://github.com/apache/commons-io/blob/rel/commons-io-2.8.0/src/main/java/org/apache/commons/io/IOUtils.java#L399-L402

https://github.com/apache/commons-io/blob/75f20dca72656225d0dc8e7c982e40caa9277d42/src/main/java/org/apache/commons/io/IOUtils.java#L465-L467

**Comment by LuciferYang** [comment]
@HeartSaVioR I found that the `@deprecated` has been removed in `commons-io` 2.11.0. I will give a pr to upgrade `commons-io`  to version 2.11.0, and revert the change of this pr to maintain old behavior(don't print error log). 

Do you think it's ok?

**Comment by HeartSaVioR** [comment]
I'm OK with it. Could you please double check there is no change on transitive dependency during upgrading before submitting PR? Thanks in advance!

**Comment by LuciferYang** [comment]
> I'm OK with it. Could you please double check there is no change on transitive dependency during upgrading before submitting PR? Thanks in advance!

OK

**Comment by dongjoon-hyun** [comment]
+1 for the direction upgrading `commons-io` and reverting this. Thanks!

**Comment by LuciferYang** [comment]
> +1 for the direction upgrading commons-io and reverting this. Thanks!

OK ~ in one ?

**Pull request closed without merging**